vita3k/packages/src/sce_utils.cpp

```diff
@@ -1101,20 +833,253 @@ std::tuple<uint64_t, SelfType> get_key_type(std::ifstream &file, const SceHeader
+            segment_infos.push_back(segment_info);
+
+            if (segment_info.plaintext == SecureBool::NO)
+                encrypted = true;
+        }
+
+        std::vector<SceSegment> scesegs;
+
+        if (encrypted)
+            scesegs = get_segments(fself.data(), sce_hdr, SCE_KEYS, app_info_hdr.sys_version, app_info_hdr.self_type, npdrmtype, klic);
+
+        EVP_CIPHER_CTX *cipher_ctx = EVP_CIPHER_CTX_new();
+        EVP_CIPHER *cipher = EVP_CIPHER_fetch(nullptr, "AES-128-CTR", nullptr);
+        int dec_len = 0;
+
+        for (uint16_t i = 0; i < elf_hdr.e_phnum; i++) {
+            int idx = 0;
+
+            if (!scesegs.empty())
+                idx = scesegs[i].idx;
+            else
+                idx = i;
+            if (elf_phdrs[idx].p_filesz == 0)
+                continue;
+
+            const int pad_len = elf_phdrs[idx].p_offset - at;
+            if (pad_len < 0)
+                LOG_ERROR("ELF p_offset Invalid");
+
+            std::vector<char> padding;
+            for (int i = 0; i < pad_len; i++) {
+                padding.push_back('\0');
+            }
+
+            elf.insert(elf.end(), padding.begin(), padding.end());
+
+            at += pad_len;
+
+            std::vector<unsigned char> dat(segment_infos[idx].size);
+            memcpy(dat.data(), &fself[segment_infos[idx].offset], segment_infos[idx].size);
+
+            std::vector<unsigned char> decrypted_data(segment_infos[idx].size);
+            if (segment_infos[idx].plaintext == SecureBool::NO) {
+                EVP_DecryptInit_ex(cipher_ctx, cipher, nullptr, reinterpret_cast<const unsigned char *>(scesegs[i].key.c_str()), reinterpret_cast<const unsigned char *>(scesegs[i].iv.c_str()));
+                EVP_CIPHER_CTX_set_padding(cipher_ctx, 0);
+                EVP_DecryptUpdate(cipher_ctx, decrypted_data.data(), &dec_len, dat.data(), segment_infos[idx].size);
+                EVP_DecryptFinal_ex(cipher_ctx, &decrypted_data[dec_len], &dec_len);
+            }
+
+            if (segment_infos[idx].compressed == SecureBool::YES) {
+                const std::string decompressed_data = decompress_segments(decrypted_data, segment_infos[idx].size);
+                segment_infos[idx].compressed = SecureBool::NO;
+                elf.insert(elf.end(), decompressed_data.begin(), decompressed_data.end());
+                at += decompressed_data.length();
+            } else {
+                elf.insert(elf.end(), decrypted_data.data(), &decrypted_data[segment_infos[idx].size]);
+                at += segment_infos[idx].size;
+            }
+        }
+
+        EVP_CIPHER_CTX_free(cipher_ctx);
+        EVP_CIPHER_free(cipher);
+
+        // Create a new self with the extracted elf
+        std::vector<uint8_t> decrypted_self(HEADER_LEN + elf.size());
+        ElfHeader ehdr = ElfHeader(elf.data());
+
+        // Create a new header
+        SCE_header hdr{};
```

Comment:
```suggestion
        SCE_header hdr{
           .magic = SCE_MAGIC,
           .version = 3,
           .sdk_type = 0xC0,
           ...
        };
```

Reply:
Oh good, i have never using add entry like that with . ^^

Reply:
hmm i have some issie for trying doing like that, probleme on last line